client/client_test.go

```diff
@@ -0,0 +1,553 @@
+	}
+
+	reconcilerObj = &reconciler{
+		ResultsChan: make(chan ObjectIdentifier, 20),
+	}
+
+	dynamicWatcher, err = New(k8sConfig, reconcilerObj, nil)
+	Expect(err).To(BeNil())
+	Expect(dynamicWatcher).NotTo(BeNil())
+
+	// Return the named return values.
+	return
+}
+
+var _ = Describe("Test the client", Ordered, func() {
+	var (
+		ctxTest        context.Context
+		cancelCtxTest  context.CancelFunc
+		dynamicWatcher DynamicWatcher
+		reconcilerObj  *reconciler
+		watched        []*corev1.Secret
+		watchedObjIDs  []ObjectIdentifier
+		watcher        *corev1.ConfigMap
+	)
+
+	BeforeAll(func() {
+		ctxTest, cancelCtxTest = context.WithCancel(ctx)
+		watcher, watched, reconcilerObj, dynamicWatcher = getDynamicWatcher(ctx)
+
+		watchedObjIDs = []ObjectIdentifier{}
+		for _, watchedObj := range watched {
+			watchedObjIDs = append(watchedObjIDs, toObjectIdentifer(watchedObj))
+		}
+
+		go func() {
+			defer GinkgoRecover()
+
+			err := dynamicWatcher.Start(ctxTest)
+			Expect(err).To(BeNil())
+		}()
+
+		<-dynamicWatcher.Started()
+	})
+
+	AfterEach(func() {
+		// Drain the test results channel.
+		for len(reconcilerObj.ResultsChan) != 0 {
+			_, ok := <-reconcilerObj.ResultsChan
+			Expect(ok).To(BeTrue())
+		}
+	})
+
+	AfterAll(func() {
+		for _, objectID := range watched {
+			err := k8sClient.CoreV1().Secrets(namespace).Delete(ctxTest, objectID.Name, metav1.DeleteOptions{})
+			if !k8serrors.IsNotFound(err) {
+				Expect(err).To(BeNil())
+			}
+		}
+
+		err := k8sClient.CoreV1().ConfigMaps(namespace).Delete(ctxTest, watcher.Name, metav1.DeleteOptions{})
+		if !k8serrors.IsNotFound(err) {
+			Expect(err).To(BeNil())
+		}
+
+		cancelCtxTest()
+	})
+
+	It("Ensures watches are not added when invalid input is provided", func() {
+		By("Verifying that an error is returned when no watched objects are provided")
+		err := dynamicWatcher.AddOrUpdateWatcher(toObjectIdentifer(watcher))
+		Expect(errors.Is(err, ErrInvalidInput)).To(BeTrue())
+
+		By("Verifying that an error is returned when an invalid watcher object is provided")
+		err = dynamicWatcher.AddOrUpdateWatcher(ObjectIdentifier{}, watchedObjIDs...)
+		Expect(errors.Is(err, ErrInvalidInput)).To(BeTrue())
+
+		By("Verifying that an error is returned when an invalid watched object is provided")
+		err = dynamicWatcher.AddOrUpdateWatcher(toObjectIdentifer(watcher), ObjectIdentifier{})
+		Expect(errors.Is(err, ErrInvalidInput)).To(BeTrue())
+
+		By("Verifying that an error is returned when the watched object does not have a CRD installed")
+		obj := ObjectIdentifier{
+			Group:     "",
+			Version:   "v1",
+			Kind:      "SuperCoolConfigMap",
+			Namespace: "test-ns",
+			Name:      "watcher",
+		}
+		err = dynamicWatcher.AddOrUpdateWatcher(toObjectIdentifer(watcher), watchedObjIDs[0], obj)
+		Expect(err).ToNot(BeNil())
+
+		By("Verifying that the previous encountered error resulted in the watch being cleaned up")
+		Eventually(dynamicWatcher.GetWatchCount, 3).Should(Equal(uint(0)))
```

Comment:
Nit: I think I prefer gomega intervals to use the string format, so I think this would be:

```suggestion
		Eventually(dynamicWatcher.GetWatchCount, "3s").Should(Equal(uint(0)))
```

I just think it's more clear, because the unit is explicitly mentioned.